pkg/machinehealthcheck/reconciler/reconciler.go

```diff
@@ -0,0 +1,70 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	anywhereCluster "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	"github.com/aws/eks-anywhere/pkg/controller/clientutil"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+	"github.com/aws/eks-anywhere/pkg/machinehealthcheck"
+)
+
+// Reconciler allows to reconcile machine health checks.
+type Reconciler struct {
+	client    client.Client
+	defaulter anywhereCluster.MachineHealthCheckDefaulter
+}
+
+// New returns a new Reconciler.
+func New(client client.Client, defaulter anywhereCluster.MachineHealthCheckDefaulter) *Reconciler {
+	return &Reconciler{
+		client:    client,
+		defaulter: defaulter,
+	}
+}
+
+// Reconcile installs machine health checks for a given cluster.
+// nolint:gocyclo
+func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *anywherev1.Cluster) error {
+	log.Info("Installing machine health checks for workload", "cluster", cluster.Name)
+
+	if cluster.Spec.MachineHealthCheck == nil {
+		mhc, err := machinehealthcheck.GetMachineHealthChecks(ctx, r.client, cluster)
+		if err != nil {
+			return err
+		}
+		if mhc != nil {
+			if mhc.Spec.NodeStartupTimeout != nil {
+				cluster.Spec.MachineHealthCheck = &anywherev1.MachineHealthCheck{
+					NodeStartupTimeout: mhc.Spec.NodeStartupTimeout,
+				}
+			}
+
+			if len(mhc.Spec.UnhealthyConditions) > 0 {
+				cluster.Spec.MachineHealthCheck = &anywherev1.MachineHealthCheck{
+					UnhealthyMachineTimeout: &mhc.Spec.UnhealthyConditions[0].Timeout,
+				}
+			}
+		}
+	}
+
+	cluster, err := r.defaulter.MachineHealthCheckDefault(ctx, cluster)
```

Comment:
this would set the defaults in case the eksa spec does not contain machine health checks.

Reply:
I'm confused. If the cluster doesn't have anything configured (existing cluster before this release), the block above will read the existing MHC and populate the cluster, that's great.

Otherwise, for all new clusters, shouldn't the webhook set the defaults? This way this reconciler only takes care of old clusters (this is a temporary thing, only required for the transition) and of creating/updating MHC.

Reply:
The reason we don't want to move the defaults to the webhook level is because we want to default values if the machine health checks in the spec are empty. If we end up defaulting in the webhooks before getting to the reconciler, there is no way for us to detect if the values were set by the controller defaults or the user.